exercises/nondet_monad_cont/descr.md

```diff
@@ -0,0 +1,522 @@
+both allow performing case analysis on a computation and distinguishing
+whether it fails or produces at least one result. `msplit` and `sols` differ
+in their return types: `msplit` produces an option, whereas `sols` produces a
+sequence. It is in fact possible to use `msplit` in the definition of `sols`;
+we have preferred to ask for a direct definition of `sols` above.
+
+**Question 2.** Implement the constructor function `at_most_once`.
+
+*Hint.* Use `msplit` to determine if the computation at hand fails or produces
+at least one result. In the first case, recall that `at_most_once fail` is
+supposed to behave like `fail`. In the second case, recall that
+`at_most_once (choose (return x) m)` is supposed to behave like
+`return x`.
+
+*Hint.* Use `delay` to ensure that the case analysis is performed when
+the computation `at_most_once m` is *executed*, not when it is *built*.
+
+**Question 3.** Implement the constructor function `interleave`.
+
+*Hint.* Again, use `delay` and `msplit` in an appropriate manner.
+Then, ask yourself how `interleave fail m2` is supposed to behave,
+and how `interleave (choose (return x1) m1) m2` is supposed
+to behave.
+
+**Question 4.** Implement the constructor function `(>>-)`.
+
+*Hint.* Again, use `delay` and `msplit` in an appropriate manner.
+Then, ask yourself how `fail >>- m2` is supposed to behave,
+and how `(choose (return x1) m1) >>- m2` is supposed
+to behave.
+
+## Notes
+
+This exercise is based on the paper
+**"Backtracking, Interleaving, and Terminating Monad Transformers"**
+[(Kiselyov, Shan, Friedman, and Sabry, 2005)](http://okmij.org/ftp/papers/LogicT.pdf).
+One of their key ideas is to propose the operation `msplit`,
+which offers a *view* of a computation as a sequence.
+Based on this operation, many operations, including `at_most_once`,
+`interleave` and `(>>-)` above, and more,
+can be defined *outside* the abstraction barrier of the type `'a m`.
+
+It is important to note that this paper uses Haskell, a lazy programming
+language, where computations are suspended by default. Thus, where it uses
+lists, we cannot use OCaml lists; we must instead use OCaml sequences.
+Furthermore, we must sometimes be careful to explicitly delay computations:
+our `delay` constructor serves this purpose.
+
+The paper goes beyond us by implementing nondeterminism not just as a monad,
+but as a **monad transformer**. Monad transformers can be composed. For
+instance, by composing the nondeterminism monad transformer and the state
+monad transformer, one can describe computations that involve both
+nondeterminism and mutable state.
+
+In this exercise, a nondeterministic computation is represented as a `compute`
+function, which expects a success continuation and a failure continuation as
+arguments. We have noted above that these continuations are analogous to the
+data constructors `Seq.Nil` and `Seq.Cons`. This is no accident: the type `'a
```

Comment:
Is the repetition of this sentence intentional?

Reply:
I think the repetition is intentional. (The second occurrence is part of the notes, which are meant to be stand-alone to some extent.)